### PR TITLE
Create VJP perf fix

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** v0.33.1 backport of PR https://github.com/PennyLaneAI/pennylane/pull/4792

**Description of the Change:** Updates VJP pipeline to favour direct matrix-vector products where possible.

**Benefits:** Improves performance for many parameter/many observable workloads

**Possible Drawbacks:**

**Related GitHub Issues:** https://github.com/PennyLaneAI/pennylane/pull/4792

### CODE REVIEW

File: pennylane/gradients/vjp.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def _all_close_to_zero(dy):
         return qml.math.allclose(dy, 0)
 
     # call this method recursively
-    return qml.math.all(qml.math.stack([_all_close_to_zero(dy_) for dy_ in dy]))
+    return all(_all_close_to_zero(dy_) for dy_ in dy)
```
Comment (trbromley) [inline pennylane/gradients/vjp.py:50]: Two questions here:
- Could you explain what is happening with this line change?
- It appears we still use `_all_close_to_zero` early on in `vjp` - why is it ok there, and why did we also have it in `compute_vjp_single`?

File: pennylane/gradients/vjp.py [hunk 2]
```diff
@@ -115,16 +115,8 @@ def compute_vjp_single(dy, jac, num=None):
...
-            # If the dy vector is zero, then the
-            # corresponding element of the VJP will be zero.
-            num_params = len(jac) if isinstance(jac, tuple) else 1
-
-            res = qml.math.convert_like(np.zeros(num_params), dy)
-            return qml.math.cast_like(res, dy)
-    except (AttributeError, TypeError):
-        pass
+    # Note: For generality, all exception type warnings are disabled.
+    # TODO: Excplictly catalogue and update raises for known types.
```
Comment (trbromley) [inline pennylane/gradients/vjp.py:119]: Is this referring to the try/except blocks? From this text, I was expecting to see us filtering out certain warnings.

File: pennylane/gradients/vjp.py [hunk 3]
```diff
@@ -136,7 +128,12 @@ def compute_vjp_single(dy, jac, num=None):
         if num == 1:
             jac = qml.math.squeeze(jac)
         jac = qml.math.reshape(jac, (-1, 1))
-        res = qml.math.tensordot(jac, dy_row, [[0], [0]])
+        try:
+            res = dy_row @ jac
+
+        except Exception:  # pylint: disable=broad-exception-caught
+            res = qml.math.tensordot(jac, dy_row, [[0], [0]])
```
Comment (trbromley) [inline pennylane/gradients/vjp.py:135]: So here we are trying the fast path and falling back to the slower tensordot if anything goes wrong?

**Review by trbromley — COMMENTED** [review]
Thanks @mlxd @albi3ro! Nothing blocking from my side but I do have a few questions.

File: pennylane/gradients/vjp.py (hunk 2 above)
Comment (mlxd) [reply pennylane/gradients/vjp.py:119]: Yes, in relation to the exceptions. For now, we can catch all, and fallback to older behaviour if any failure appears (majority of cases will work fine within the try block). For all others, we'd need a list of the potential failures, so catching all and avoiding pylint raising a warning is sufficient for now.

File: pennylane/gradients/vjp.py (hunk 3 above)
Comment (mlxd) [reply pennylane/gradients/vjp.py]: Yes. This gave us seconds fewer of runtime in situations where the first path was able to complete. We can improve on this later, but this for now was the cheapest way to get a more improved runtime within the slow-path.

File: pennylane/gradients/vjp.py (hunk 1 above)
Comment (mlxd) [reply pennylane/gradients/vjp.py:50]: Since this is returning a true/false only, we don't need to stack etc on the returned output. Also `all` with a generator exits early, so it will stop evaluating the outputs if we get a non close to zero component, reducing the runtime overall.

Keeping this for `vjp` allows us to preserve the current behaviour of reduced executions for hardware devices.

File: pennylane/gradients/vjp.py (hunk 2 above)
Comment (trbromley) [reply pennylane/gradients/vjp.py:119]: Got it! This is very optional but could the following be a good explanation:
```suggestion
    # Note: The following uses try/except blocks to attempt a fast
    # dispatch before falling back to a slower method. For
    # simplicity, all exceptions are handled together.
    # TODO: Explicitly catalogue and update raises for known types.
```

File: pennylane/gradients/vjp.py (hunk 1 above)
Comment (trbromley) [reply pennylane/gradients/vjp.py:50]: > Keeping this for vjp allows us to preserve the current behaviour of reduced executions for hardware devices.

Not a blocker for merging, but I'm still confused here. It seems like `vjp` will check for `dy` being close to zero multiple times - once at this point and once again in the `processing_fn`. The latter has been disabled in this PR, but I don't understand why it was needed in the first place if we're doing the check here. Is it a subtlety with forward and backward passes?

File: pennylane/gradients/vjp.py [hunk 4]
```diff
@@ -136,7 +128,12 @@ def compute_vjp_single(dy, jac, num=None):
         if num == 1:
             jac = qml.math.squeeze(jac)
         jac = qml.math.reshape(jac, (-1, 1))
-        res = qml.math.tensordot(jac, dy_row, [[0], [0]])
+        try:
```
Comment (rmoyard) [inline pennylane/gradients/vjp.py:131]: I know for sure that tensorflow does not support `@`, do you know if that the only case where it would fail for the this one (compute_vjp_single)?

File: pennylane/gradients/vjp.py [hunk 5]
```diff
@@ -202,14 +206,34 @@ def compute_vjp_multi(dy, jac, num=None):
...
     else:
-        res = []
-        for d, j_ in zip(dy, jac):
-            sub_res = []
-            for j in j_:
-                sub_res.append(qml.math.squeeze(compute_vjp_single(d, j, num=num)))
-            res.append(sub_res)
-        res = qml.math.stack([qml.math.stack(r) for r in res])
-        res = qml.math.sum(res, axis=0)
+        try:
```
Comment (rmoyard) [inline pennylane/gradients/vjp.py:209]: Here I think it is failing when mixing different return type with different shapes. I am not sure if they are other cases where it is failing, maybe specific to the interface?

File: pennylane/gradients/vjp.py [hunk 6]
```diff
@@ -202,14 +206,34 @@ def compute_vjp_multi(dy, jac, num=None):
...
+            dy_interface = qml.math.get_interface(dy[0])
+            # dy  -> (i,j)      observables, entries per observable
+            # jac -> (i,k,j)    observables, parameters, entries per observable
+            # Contractions over observables and entries per observable
+            dy_shape = qml.math.shape(dy)
+            if len(dy_shape) > 1:  # multiple values exist per observable output
+                return qml.math.array(qml.math.einsum("ij,i...j", dy, jac), like=dy[0])
+
+            if dy_interface == "tensorflow":
```
Comment (rmoyard) [inline pennylane/gradients/vjp.py:218]: That only covers single measurement hessian with tf right?

**Review by rmoyard — COMMENTED** [review]
Thanks for the changes! It would be nice to have the list of cases where the it becomes slower because of the fallback. What I can identify is the following:

The performance is better:
1) Single measurement (not tensorflow) with parameters of the same shape.
2) Multiple measurements with same shapes and with parameters of the same shape.

The performance is worse:
1) Parameters of the different shapes.
2) Tensorflow in general takes a hit because it does not support `@` for mulitplication
3) Mixing measurements with different shapes is slower because of the fallback in the `compute_vjp_multi`.

In the future we should come back to dispatch the cases properly.

**Comment by timmysilv** [comment]
try-except generally hurts performance. I made a table of when we need which try-excepts so we can improve this in the future:

|                      | non-TF | TF        |
| ------------ | ------- | ------- |
| non-ragged | none    | both     |
| ragged         | multi    | both     |

**Review by rmoyard — APPROVED** [review]
Thanks! We made a small table of what cases need what try expect and it also show which one are better and which one become worse. It will be easier to come back to this issue.

**Review by albi3ro — APPROVED** [review]
Performance improvements for the vast majority of hpc cases we care about, at the cost of a penality for edge cases and tensorflow. I find that to be an acceptable tradeoff.

While this code could probably always use more fine-tuning, this makes accomplishes exactly what it needs to accomplish,
